<h1>Uma introdução à geolocalização ECEF com Python</h1>

Latitude, longitude e altitude são nomes comuns quando se trata de sistemas de geolocalização. Poucas pessoas conhecem outros sistemas como o ECEF. Neste post conheceremos esse sistema pouco comum e aprenderemos como calcular coordenadas nele. 

Este post é voltado para alunos de ensino superior que possuem pouca ou nenhuma noção de programação com Python. Recomendado para professores e alunos que estão cursando Geometria Analítica. 

<h3>Sumário</h3>

- Sistma Latitude, Longitude e Altitude (LLA)
- Sistema Terra-Centro Terra-Fixa (ECEF)
- A matemática por transformação de LLA para ECEF
- Obtendo coordenadas LLA de um ponto
- De LLA para ECEF com Python: a ferramenta
- De LLA para ECEF com Python: o código (em 4 linhas)
- Aplicação: A área do Palácio da Alvorada!

<h3>Sistema LLA</h3>

No sistema Latitude-Longitude-Altitude, ou simplesmente LLA, todo ponto $P$ na terra é representado por três coordenadas: latitude $\lambda$, longitude $\phi$ e altitude $h$. Cada uma dessas coordenadas são definidas a seguir:

- $\lambda$: é a medida do ângulo entre o plano do equador e o ponto P. Ao norte essa medida atinge no máximo 90º e ao sul no mínimo -90º. Sendo assim $\lambda \in [-90,90]$.
- $\phi$: é a medida do ângulo formado pelo plano que contem o Meridiano de Greenwich e o ponto $P$. A leste e a Oeste, essa medida, também por convenção, atinge no máximo 180º.
- $h$: é a altura em relação ao nível do mar local $N$, isto é, para o cálculo de $h$, mede-se do nível do mar nas proximidades até o ponto $P$.

<h3>Sistema ECEF</h3>

No sistema ECEF, um ponto $P$ é representado por coordenadas cartesianas $(x,y,z)$. O referencial tridimensional é considerado o seguinte:

- origem: Centro do planeta terra
- eixo x: Semi-reta da origem até o ponto de interseção do Meridiano de Greenwich com o Equador.
- eixo y: No plano contendo a linha do equador move-se o eixo x 90º para o leste.
- eixo z: Do polo sul para o polo norte

<h3>A matemática da transformação de LLA para ECEF</h3>

Dados as coordenadas $\phi$, $\lambda$ e $h$ de um ponto $P$, para determinar as coordenadas $(x,y,z)$ ECEF do ponto $P$ utiliza-se a seguinte forma:

$$\left(\begin{array}{c}x \\ y \\ z \end{array}\right) = \left(\begin{array}{c} (N+h)cos\phi cos\lambda \\ (N+h)cos\phi cos\lambda \\ (N(1-e^2)+h)sen\phi \end{array}\right)$$

<h3>Obtendo coordenadas LLA de um ponto</h3>

Existem vários sites que oferecem esse serviço. Para latitude e longitude, o mais comum é o Google Maps. Consideraremos o ponto $P$ sendo o palácio da Alvorada, residência do presidente do Brasil.

<img src="mapPalacioAlvorada.png">

No google Maps, após digitar o endereço procurado na busca, o endereço mostrado pela barra de navegação do seu navegador revela a latitude a a longitude daquele local.

https://www.google.com.br/maps/place/Palacio+da+Alvorada+-+Brasília,+DF,+70297-400/@-15.7899211,-47.9038544,12z/

Neste exemplo, a latitude é -15.7899 e a longitude é -47.9038.

O site Elevation finder https://www.freemaptools.com/elevation-finder.htm mostra pra você qual a altura dado um endereço.

<img src="elevationPalacioAlvorada.png">

<h3>De LLA para EFEC na prática</h3> 

O site do departamento de oceanografia da Naval PostGraduate School https://www.nps.edu/ fornece um um conversor simples e intuitivo https://www.oc.nps.edu/oc2902w/coord/llhxyz.htm 

Utilizaremos as coordenadas do Palácio da Alvorada para esse exemplo.

<img src="LLAtoECEF.png">

<h3>De LLA para ECEF com Python: a ferramenta</h3>

Utilizaremos Python para determinar as coordenadas $(x,y,z)$ encontradas acima. A escolha de Python se deve pelo fato de ser uma linguagem de programação recomendada para iniciantes. Para executarmos nossso código em Python utilizaremos o Jupyter Colabore do Google. Assim, não precisaremos instalar nada no computador.

Entre no site: https://colab.research.google.com/notebooks/welcome.ipynb#recent=true

Vá em "File" e depois em "New Python 3 Notebook". Logue com a conta do gmail.

Se tudo correu bem aparecerá a seguinte tela:

<img src="colaboreInicio.png">

O colabore funciona como uma calculadora. Digitamos uma "conta" e ele retorna o resultado depois de apertarmos "play".

<img src="conta.png">

O exemplo acima é apenas uma simples e humilde ilustração de como o Jupyter colabore funciona. Na verdade, ele pode calcular desde integrais até a probabilidade da ação de uma grande empresa cair. Utilizaremos o Jupyter colabore para transformar as coordenadas LLA para ECEF. Utilizar um programa que faça isso é importante para automatizar os processos e automatizar um processo é importante para diminuir custos e tempo.

Para criar uma nova célula e assim poder executar mais comandos, use o botão "Code" mostrado na imagem abaixo:

<img src="code.png">

<h3>Faça isso você mesmo: O código</h3>

Muitos programadores criam bibliotecas com "contas" mais elaboradas. Para usar alguma dessas "contas" temos que seguir os três passos: 

- instalar a biblioteca caso já não a tenha instalado uma vez
- importar a função desejada da biblioteca
- chamar a função de acordo com as instruções desta

Utilizando "hashtags" antes de um texto faz com que o Python não leia aquela linha. Logo utiliza-se "hashtags" para fazer comentários no código. Dessa forma, as linhas com "hashtags" a seguir podem ser totalmente ignoradas.

In [1]:
# instalando a biblioteca
!pip install pymap3d

<img src="instalandoPymap.png">

In [2]:
# importando a função desejada da biblioteca
from pymap3d import geodetic2ecef

<img src="importandoFuncao.png">

In [3]:
# chamando a função de acordo com as instruções desta
x, y, z = geodetic2ecef( -15.79263, -47.82212, 1014)
print(x, y, z)

4122527.1043634596 -4550042.799842703 -1724934.4312999998


<img src="chamandoFuncao.png">

Observe que nesse caso, as coordenadas ECEF são dadas em metros.

<h3>A área do Palácio da Alvorada</h3>

Como aplicação estimaremos a área, incluindo o espaço verde, do Palácio da Alvorada.

Primeiro, notamos que a área é aproximadamente um triângulo.

<img src="triangulo.png">

As coordenadas LLA em metros dos pontos A, B e C, obtidas através do site "freemaptools", são aproximadamente:

A = (-15.78840, -407.82532, -9999.0)

B = (-15.79550, -407.82558, -9999.0)

C = (-15.79197, -407.81884, -9999.0)

Convertendo para ECEF temos:

In [4]:
x, y, z = geodetic2ecef( -15.78840, -407.82532, -9999.0)
print("A = ({}, {}, {})".format(x, y, z))

A = (4115243.4089937783, -4542513.618083791, -1721487.4690041905)


In [5]:
x, y, z = geodetic2ecef( -15.79550, -407.82558, -9999.0)
print("B = ({}, {}, {})".format(x, y, z))

B = (4115079.4695950192, -4542374.083886003, -1722242.2844763028)


In [6]:
x, y, z = geodetic2ecef( -15.79197, -407.81884, -9999.0)
print("C = ({}, {}, {})".format(x, y, z))

C = (4115685.0605882085, -4541968.633198852, -1721867.006226004)


In [7]:
import numpy as np

# Os pontos A, B e C devem ser declarados como segue
A = np.array(
    (4115243.4089937783, -4542513.618083791, -1721487.4690041905)
)
B = np.array(
    (4115079.4695950192, -4542374.083886003, -1722242.2844763028)
)
C = np.array(
    (4115685.0605882085, -4541968.633198852, -1721867.006226004)
)

# fazendo o produto vetorial de AB com AC
ABxAC = np.cross(B-A, C-A)

# calculando a área do triângulo ABC
AREA = np.linalg.norm(ABxAC)/2

# Imprimindo o resultado
print(AREA,"metros quadrados")

277369.1244671427 metros quadrados


A área é de aproximadamente 277369 metros quadrados!

<h3>Conclusão</h3>

Neste post aprendemos:

- O que é o sistema LLA
- O que é o sistema ECEF
- Como obter coordenadas LLA
- Como converter LLA para ECEF com Python
- Como calcular área utilizando ECEF

Até a próxima! \o